File: Rate_limiter/rateLimiter.go
package main

import (
	"fmt"
	"math"
	"time"
)

type ID string
type TimeInMillis float64
type Limit float64
type DecoratedFunc func(ID) float64
type Decorator func(DecoratedFunc) DecoratedFunc

func createRateLimiter(limit Limit) Decorator {
	lastCalled := make(map[ID]TimeInMillis)

	return func(f DecoratedFunc) DecoratedFunc {
		return func(id ID) float64 {
			currentTime := TimeInMillis(time.Now().UnixMilli())

			// if the ID has not been called before
			if _, ok := lastCalled[id]; !ok {
				lastCalled[id] = currentTime
				return f(id)
			}

			// if the ID has been called before
			timeSinceLastCall := currentTime - lastCalled[id]
			if timeSinceLastCall < TimeInMillis(1/limit) {
				fmt.Println("Rate limit exceeded.")
				return 0.0 // assuming float64 return type
			}
			lastCalled[id] = currentTime
			return f(id)
		}
	}
}

func returnPi(id ID) float64 {
	return math.Pi
}

File: Rate_limiter/rate_limiter.py
import time
import math


def rate_limiter(limit):
    def decorator(func):
        last_called = {}

        def wrapper(id):
            current_time = (
                time.time() * 1000
            )  # the * 1000 is to convert to milliseconds

            # if the ID has not been called before
            if id not in last_called:
                last_called[id] = current_time
                return func(id)

            # if the ID has been called before
            time_since_last_call = current_time - last_called[id]
            if time_since_last_call < 1 / limit:
                return "Rate limit exceeded. Please try again later."
            last_called[id] = current_time
            return func(id)

        return wrapper

    return decorator


# Limit is 1 call within 200 milliseconds
@rate_limiter(1 / 200)
def return_pi(id):
    return math.pi
